File: app/src/main/java/repo/dao/CategoryDao.java
package repo.dao;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

import model.Category;

@Dao
public interface CategoryDao {

    @Insert(entity = Category.class, onConflict = OnConflictStrategy.REPLACE)
    void create(Category category);

    @Update(entity = Category.class, onConflict = OnConflictStrategy.REPLACE)
    void update(Category category);

    @Delete(entity = Category.class)
    void delete(Category category);


    @Query("SELECT * FROM categories ORDER BY id DESC;")
    List<Category> getAllCategories();

    @Query("SELECT * FROM categories WHERE id == :categoryID;")
    Category getCategory(long categoryID);

    @Query("DELETE FROM categories;")
    void deleteAllCategories();

    @Query("SELECT COUNT(*) FROM categories;")
    long getCategoriesCount();

}

File: app/src/main/java/repo/dao/CategoryDao.kt
package repo.dao

import androidx.room.*
import kotlinx.coroutines.flow.Flow
import model.Category

@Dao
interface CategoryDao {

    @Insert(entity = Category::class, onConflict = OnConflictStrategy.REPLACE)
    suspend fun create(category: Category?)

    @Update(entity = Category::class, onConflict = OnConflictStrategy.REPLACE)
    suspend fun update(category: Category?)

    @Delete(entity = Category::class)
    suspend fun delete(category: Category?)

    @Query("SELECT * FROM categories ORDER BY id DESC;")
    fun getAllCategories(): Flow<List<Category>?>

    @Query("SELECT * FROM categories WHERE id == :categoryID;")
    suspend fun getCategory(categoryID: Long): Category

    @Query("DELETE FROM categories;")
    suspend fun deleteAllCategories()

    @Query("UPDATE todos SET category_id = 0, category = null;")
    suspend fun clearAllCategories()

    @Query("UPDATE todos SET category_id = 0, category = null WHERE category_id = :categoryID;")
    suspend fun clearSingleCategory(categoryID: Long)

    @Query("SELECT COUNT(*) FROM categories;")
    suspend fun getCategoriesCount(): Long

    @Query("UPDATE todos SET category = :categoryName WHERE category_id = :categoryID;")
    suspend fun editTodoCategory(categoryID: Long, categoryName: String?)
}